core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -815,6 +815,23 @@ public Optional<TableLayout> getNewTableLayout(Session session, String catalogNa
                 .map(layout -> new TableLayout(catalogHandle, transactionHandle, layout));
     }
 
+    @Override
+    public ConnectorTableMetadata getNewTableMetadata(Session session, String catalogName, QualifiedObjectName table, List<ColumnMetadata> columns, Map<String, Object> properties, Optional<String> comment)
+    {
+        CatalogMetadata catalogMetadata = getRequiredCatalogMetadata(session, catalogName);
+        CatalogHandle catalogHandle = catalogMetadata.getCatalogHandle(session, table);
+        ConnectorMetadata metadata = catalogMetadata.getMetadataFor(session, catalogHandle);
+        ConnectorSession connectorSession = session.toConnectorSession(catalogHandle);
+        ConnectorTableMetadata newTableMetadata = metadata.getNewTableMetadata(
+                connectorSession,
+                table.asSchemaTableName(),
+                columns,
+                properties,
+                comment);
+        checkState(newTableMetadata.getColumns().size() == columns.size(), "Table and query column count doesn't match");
```

Comment:
What query is the message talking about?

Reply:
I took this terminology from the `INSERT` vs table but maybe this is slightly flawed in the case of a CTAS.

So the query is the CTAS query, the table is the effective storage table.